### ISSUE TITLE
[4.0] Installer - username validation

### ISSUE THREAD

**Issue by astridx** [joomla/joomla-cms#19301]
### Steps to reproduce the issue
Install Joomla version 4 and call the super user 
**admin&**
admin with an following ampersand

### Expected result
The user should be saved with the name 
**admin&**
admin with an following ampersand or there should be a warning or the the user should not be saved

### Actual result
The user is saved with the user name admin. The ampersand is cut off.

### System information (as much as possible)
Current Joomla! 4.0.0-dev Development [ Amani ]

**Comment by brianteeman** [comment]
The same problem exists for passwords - reported elsewhere just add this to that issue.

**Comment by ghost** [comment]
#19251, #19204

**Comment by astridx** [comment]
@brianteeman 
I am not sure if it is the same issue. We do not allow user names with an ampersand up to now. If you create an user with an ampersand in the user name via user manager, you see a warning.
This warning is because of JOOMLA/media/system/js/fields/validate.min.js 

So we should decide if we want allow user names with special characters (RegExp("[\<|\>|\"|\'|\%|\;|\(|\)|\&]", "i")).

**Comment by brianteeman** [comment]
Testing this again 
Creating a user in the backend fails with the correct error message

Creating a user at installation allows you to create a user with an & and it is stored in the db with an &
However that user cannot log in

So for me the error is in the installer and it not using the same filtering as the rest of joomla

**Comment by brianteeman** [comment]
In fact looking closer the installer uses no filtering at all

**Comment by brianteeman** [comment]
The problem is that users are validated in the installer with JOOMLA/media/system/js/fields/validate.min.js
`        const regex = new RegExp('[<|>|"|\'|%|;|(|)|&]', 'i');
`
But in the admin they are validated with a completely different regex
https://github.com/joomla/joomla-cms/blob/064cf17b64698cc7f999e40c4f38803f11e41f2b/libraries/src/Table/User.php#L206

It looks like the email is validated differently as well

**Comment by dgrammatiko** [comment]
> The problem is that users are validated in the installer with JOOMLA/media/system/js/fields/validate.min.js
> const regex = new RegExp('[<|>|"|\'|%|;|(|)|&]', 'i');
> But in the admin they are validated with a completely different regex

Just use the HTML5 `pattern` in the XML and use the same regex. That should fix the inconsistency (assuming that the pattern is working correctly otherwise blame me and I'll get it fixed)

**Comment by brianteeman** [comment]
I can do that (and agree) but then we have multiple unused regex in multiple places 

Remember I wanted to do https://github.com/joomla/joomla-cms/issues/18995

**Comment by brianteeman** [comment]
Plus the error message from an html5 pattern isn't very good as it doesnt tell you why it fails

<img width="300" alt="chrome_2018-08-12_21-18-25" src="https://user-images.githubusercontent.com/1296369/44006056-57012714-9e75-11e8-9431-d9e618850f94.png">

I assume though that it is automatically translated according to the browser language

**Comment by ggppdk** [comment]
> Just use the HTML5 pattern in the XML and use the same regex

> Plus the error message from an html5 pattern isn't very good as it doesnt tell you why it fails

text field layout now has ...

```php
$attributes = array(
	...
	strlen($hint) ? 'placeholder="' . htmlspecialchars($hint, ENT_COMPAT, 'UTF-8') . '"' : '',
	...
	!empty($pattern) ? 'pattern="' . $pattern . '"' : '',
);
?>
```

We could try adding support 1 new attribute to the XML files
`error_hint="..."`

```php
$attributes = array(
	...
	strlen($hint) ? 'placeholder="' . htmlspecialchars($hint, ENT_COMPAT, 'UTF-8') . '"' : '',
	...
	!empty($pattern) ? 'pattern="' . $pattern . '"' : '',
	!empty($error_hint) ? 'oninvalid="setCustomValidity(\'' . htmlspecialchars($error_hint, ENT_QUOTES, 'UTF-8') . '\');" ' : '',
	!empty($error_hint) ? 'oninput="setCustomValidity(\'\');" ' : '',
);
?>
```

**Comment by ggppdk** [comment]
I see a new attribute for XML definition of text field already exists in J4 and it is called ?
'validationtext'

thus the above can be used ?

**Comment by ggppdk** [comment]
Thus i suggest adding attributes 
pattern 
validationtext

to the password field like they exist in the text field

[EDIT]
and to any other fields that can possibly use them (if there are any other such fields)

**Comment by dgrammatiko** [comment]
@ggppdk I've already implemented the `validationtext ` because I thought it will be handy. I guess I should have implemented one field as a showcase as it's now pretty hard to figure out without reading the code. Ooops sorry...

**Comment by brianteeman** [comment]
@dgrammatiko give me an example and I will see what I can do. This is especially important for accessibility as it will allow us to have inline error messages instead of having them out of context in an alert

**Comment by dgrammatiko** [comment]
@brianteeman the xml should be like (you'll need to use the right regex though):
<img width="628" alt="screenshot 2018-08-12 at 23 29 23" src="https://user-images.githubusercontent.com/3889375/44006627-b75415f6-9e87-11e8-958c-90a2ff091a46.png">

and then it should just work:
<img width="598" alt="screenshot 2018-08-12 at 23 29 12" src="https://user-images.githubusercontent.com/3889375/44006631-c4ac1820-9e87-11e8-8c60-f0b2e52673b3.png">

**Comment by brianteeman** [comment]
the regex is no problem ;)

as for the validationtext - does it support Jtext?

**Comment by dgrammatiko** [comment]
@brianteeman that part was never implemented. in `libraries/src/Form/FormField.php` line 1002 change it to `'validationtext' => Text::_($this->validationtext),`. That should be enough to introduce the translations
<img width="825" alt="screenshot 2018-08-12 at 23 51 27" src="https://user-images.githubusercontent.com/3889375/44006815-dfaa36d6-9e8a-11e8-96ac-5430e397c761.png">

**Comment by brianteeman** [comment]
Ok well I can see what I am doing tomorrow

**Comment by Quy** [comment]
Fixed with PR #23918 in J3.

**Comment by brianteeman** [comment]
@quy that might not fix the issue here as the installer for j4 is almost completely different code

**Comment by alikon** [comment]
imho #23918 is not enough we lack the client side validation
but in the meantime test the server side one :smile: #24024

**Comment by dgrammatiko** [comment]
@alikon @brianteeman the front end part was already fixed in 
https://github.com/joomla/joomla-cms/pull/21584 but then that PR was closed. FWIW my intension when I did patch the validation.js to deal with the pattern attribute was to move to HTML5 validation and also have one regex per field (php or js). Obviously converting all the forms is kinda big task but honestly I hope someone will do it for J4. Brian already showcased what needs to be done, it's not hard but it's a big task

**Comment by Quy** [comment]
Fixed with #23918 that has ported to J4.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19301">issues.joomla.org/tracker/joomla-cms/19301</a>.</sub>
